relayer/client.go

```diff
@@ -169,7 +169,10 @@ func CreateClient(ctx context.Context, src, dst *Chain, srcUpdateHeader, dstUpda
 
 				if !success {
 					src.LogFailedTx(res, err, msgs)
-					return fmt.Errorf("tx failed on chain{%s}: %s", src.ChainID(), res.Data)
+					if res != nil {
```

Comment:
I'm not sure about this change. If `err` was nil, and `success` was false, _and_ `res` was nil, is it appropriate to continue to the `return err` on L177 (which is effectively `return nil` anyway)?

As it stands right now, the cosmos provider's `SendMessages` implementation cannot return a nil relayer response with a nil error. So this won't change anything for that implementation, but it seems like the intent from CreateClient is that it should only return nil on success. I don't know what a nil response and a nil error from SendMessages is supposed to indicate. We haven't documented it in the ChainProvider interface either.

Reply:
You're right we would still want to be returning an error here. 

I don't think `SendMessages` should be returning a nil response and nil error ever. This seems to have been some weird edge case which I haven't been able to reproduce since this morning.

Logs from when I encountered the panic.
```
2022-04-14T15:54:35.405256Z     debug   No client found on source chain tracking the state of counterparty chain        {"src_chain_id": "cosmoshub-4", "dst_chain_id": "osmosis-1"}
2022-04-14T15:54:35.405341Z     info    Failed sending transaction      {"chain_id": "osmosis-1", "msg-0": {"msg_json": "{\"client_state\":{\"chain_id\":\"cosmoshub-4\",\"trust_level\":{\"numerator\":\"1\",\"denominator\":\"3\"},\"trusting_period\":\"1540800000000000\",\"unbonding_period\":\"1814400000000000\",\"max_clock_drift\":\"600000000000\",\"frozen_height\":{},\"latest_height\":{\"revision_number\":\"4\",\"revision_height\":\"10116294\"},\"proof_specs\":[{\"leaf_spec\":{\"hash\":1,\"prehash_value\":1,\"length\":1,\"prefix\":\"AA==\"},\"inner_spec\":{\"child_order\":[0,1],\"child_size\":33,\"min_prefix_length\":4,\"max_prefix_length\":12,\"hash\":1}},{\"leaf_spec\":{\"hash\":1,\"prehash_value\":1,\"length\":1,\"prefix\":\"AA==\"},\"inner_spec\":{\"child_order\":[0,1],\"child_size\":32,\"min_prefix_length\":1,\"max_prefix_length\":1,\"hash\":1}}],\"upgrade_path\":[\"upgrade\",\"upgradedIBCState\"],\"allow_update_after_expiry\":true,\"allow_update_after_misbehaviour\":true},\"consensus_state\":{\"timestamp\":\"2022-04-14T15:54:01.570761059Z\",\"root\":{\"hash\":\"YgqMyJV12scwLZXdNOgI76E7x9tdjIs8dZDxSGr4+Qg=\"},\"next_validators_hash\":\"A2F52783C60438185063980C09AE849F00EF5FF1996B2E941A87BC1CE39250AF\"},\"signer\":\"osmo1kn4tkqezr3c7zc43lsu5r4p2l2qqf4mpevqzt5\"}"}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x30 pc=0x103443ccc]
```